File: Frontend/src/types/task.ts
export type BoardColumn = "backlog" | "todo" | "in_progress" | "done";
export type TaskPriority = "low" | "medium" | "high" | "urgent";

export interface EchoTask {
  id: number;
  task: string;
  deadline: string | null;
  status: string;
  note_id?: number;
  note_filename?: string | null;
  priority?: string;
  assignee?: string | null;
  board_column?: string;
  position_x?: number | null;
  position_y?: number | null;
  completed_at?: string | null;
  created_at?: string;
}

export interface TaskAnalyticsSummary {
  total_tasks: number;
  completed_tasks: number;
  completion_rate: number;
  active_tasks: number;
  overdue_count: number;
  by_assignee: Record<string, { total: number; completed: number }>;
  by_priority: Record<string, number>;
  by_column: Record<string, number>;
  completed_last_7_days: number;
  created_last_7_days: number;
  avg_days_to_complete: number | null;
}

export type TaskPatch = Partial<{
  status: string;
  board_column: BoardColumn;
  priority: TaskPriority;
  assignee: string | null;
  deadline: string | null;
  task: string;
  position_x: number;
  position_y: number;
}>;

File: Frontend/meeting-backend/schemas/task.py
from pydantic import BaseModel, Field
from datetime import datetime
from typing import Optional, Literal

BoardColumn = Literal["backlog", "todo", "in_progress", "done"]
Priority = Literal["low", "medium", "high", "urgent"]


class TaskCreate(BaseModel):
    task: str = Field(..., min_length=1)
    note_id: Optional[int] = None
    deadline: Optional[str] = None
    priority: Priority = "medium"
    assignee: Optional[str] = None
    board_column: BoardColumn = "todo"


class TaskResponse(BaseModel):
    id: int
    note_id: int
    note_filename: Optional[str] = None
    task: str
    deadline: Optional[str]
    status: str
    priority: str = "medium"
    assignee: Optional[str] = None
    board_column: str = "todo"
    position_x: Optional[float] = None
    position_y: Optional[float] = None
    completed_at: Optional[datetime] = None
    created_at: datetime

    class Config:
        from_attributes = True


class TaskUpdate(BaseModel):
    """Partial update: Kanban moves, completion, planner fields, canvas position."""

    status: Optional[str] = None
    board_column: Optional[BoardColumn] = None
    priority: Optional[Priority] = None
    assignee: Optional[str] = None
    deadline: Optional[str] = None
    task: Optional[str] = None
    position_x: Optional[float] = None
    position_y: Optional[float] = None


class TaskAnalyticsSummary(BaseModel):
    total_tasks: int
    completed_tasks: int
    completion_rate: float
    active_tasks: int
    overdue_count: int
    by_assignee: dict[str, dict[str, int]]
    by_priority: dict[str, int]
    by_column: dict[str, int]
    completed_last_7_days: int
    created_last_7_days: int
    avg_days_to_complete: Optional[float]
